codec-socks/src/main/java/io/netty/handler/codec/socksx/v5/Socks5ClientEncoder.java

```diff
@@ -0,0 +1,106 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.netty.handler.codec.socksx.v5;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandler.Sharable;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.EncoderException;
+import io.netty.handler.codec.MessageToByteEncoder;
+import io.netty.util.CharsetUtil;
+import io.netty.util.internal.StringUtil;
+
+import java.util.List;
+
+/**
+ * Encodes a client-side {@link Socks5Message} into a {@link ByteBuf}.
+ */
+@Sharable
+public class Socks5ClientEncoder extends MessageToByteEncoder<Socks5Message> {
+
+    public static final Socks5ClientEncoder DEFAULT = new Socks5ClientEncoder();
+
+    private final Socks5AddressEncoder addressEncoder;
+
+    /**
+     * Creates a new instance with the default {@link Socks5AddressEncoder}.
+     */
+    protected Socks5ClientEncoder() {
+        this(Socks5AddressEncoder.DEFAULT);
+    }
+
+    /**
+     * Creates a new instance with the specified {@link Socks5AddressEncoder}.
+     */
+    public Socks5ClientEncoder(Socks5AddressEncoder addressEncoder) {
+        if (addressEncoder == null) {
+            throw new NullPointerException("addressEncoder");
+        }
+
+        this.addressEncoder = addressEncoder;
+    }
+
+    /**
+     * Returns the {@link Socks5AddressEncoder} of this encoder.
+     */
+    protected final Socks5AddressEncoder addressEncoder() {
+        return addressEncoder;
+    }
+
+    @Override
+    protected void encode(ChannelHandlerContext ctx, Socks5Message msg, ByteBuf out) throws Exception {
+        if (msg instanceof Socks5InitialRequest) {
+            encodeAuthMethodRequest((Socks5InitialRequest) msg, out);
+        } else if (msg instanceof Socks5PasswordAuthRequest) {
+            encodePasswordAuthRequest((Socks5PasswordAuthRequest) msg, out);
+        } else if (msg instanceof Socks5CommandRequest) {
+            encodeCommandRequest((Socks5CommandRequest) msg, out);
+        } else {
+            throw new EncoderException("unsupported message type: " + StringUtil.simpleClassName(msg));
+        }
+    }
+
+    private static void encodeAuthMethodRequest(Socks5InitialRequest msg, ByteBuf out) {
+        out.writeByte(msg.version().byteValue());
+        final List<Socks5AuthMethod> authMethods = msg.authMethods();
+        out.writeByte(authMethods.size());
+        for (Socks5AuthMethod a: authMethods) {
+            out.writeByte(a.byteValue());
+        }
+    }
+
+    private static void encodePasswordAuthRequest(Socks5PasswordAuthRequest msg, ByteBuf out) {
+        out.writeByte(0x01);
+        byte[] username = msg.username().getBytes(CharsetUtil.US_ASCII);
+        byte[] password = msg.password().getBytes(CharsetUtil.US_ASCII);
+        out.writeByte(username.length);
+        out.writeBytes(username);
+        out.writeByte(password.length);
+        out.writeBytes(password);
```

Comment:
we  could use ByteBufUtil.writeAscii(...) here and just use msg.username().length() etc for the length